[quant] Add quantized Embedding module

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44217 [quant] Move EmbeddingBag eager quantization to static
* **#44208 [quant] Add quantized Embedding module**
* #44207 [quant] Support quantization of embedding lookup operators

Summary:
Add quantized module in static quantization namespace. Embedding
quantization requires only weights to be quantized so it is static.
Internally it calls the embedding_bag_byte op with the offsets set corresponding to the
indices.

Future PR will move EmbeddingBag quantization from dynamic to static as well.

Test Plan:
python test/test_quantization.py test_embedding_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23547384](https://our.internmc.facebook.com/intern/diff/D23547384)

supriyar committed:
[quant] Add quantized Embedding module

Summary:
Add quantized module in static quantization namespace. Embedding
quantization requires only weights to be quantized so it is static.
Internally it calls the embedding_bag_byte op with the offsets set corresponding to the
indices.

Future PR will move EmbeddingBag quantization from dynamic to static as well.

Test Plan:
python test/test_quantization.py test_embedding_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quantized Embedding module"


Summary:
Add quantized module in static quantization namespace. Embedding
quantization requires only weights to be quantized so it is static.
Internally it calls the embedding_bag_byte op with the offsets set corresponding to the
indices.

Future PR will move EmbeddingBag quantization from dynamic to static as well.

Test Plan:
python test/test_quantization.py test_embedding_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quantized Embedding module"


Summary:
Add quantized module in static quantization namespace. Embedding
quantization requires only weights to be quantized so it is static.
Internally it calls the embedding_bag_byte op with the offsets set corresponding to the
indices.

Future PR will move EmbeddingBag quantization from dynamic to static as well.

Test Plan:
python test/test_quantization.py test_embedding_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23547384](https://our.internmc.facebook.com/intern/diff/D23547384)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantized/modules/embedding.py:
```diff
@@ -0,0 +1,167 @@
+        self.dtype = dtype
+        if self.dtype == torch.quint8:
+            scales = torch.ones(num_embeddings, dtype=torch.float)
+            zero_points = torch.ones(num_embeddings, dtype=torch.float)
```
can we add a comment on why this is initialized to one?

vkuzo commented on torch/nn/quantized/modules/embedding.py:
```diff
@@ -0,0 +1,167 @@
+        self.sparse = sparse
+
+        if _weight is None:
+            scales = torch.ones(num_embeddings, dtype=torch.float)
+            zero_points = torch.ones(num_embeddings, dtype=torch.float)
```
seems like this is the same code as in packed params, perhaps we can only do it once?

raghuramank100 commented on torch/nn/quantized/modules/embedding.py:
```diff
@@ -0,0 +1,167 @@
+    def set_weight(self, weight):
+        # type: (torch.Tensor) -> None
+        if self.dtype == torch.quint8:
+            self._packed_weight = torch.ops.quantized.embedding_bag_prepack(weight)
```
Do we support per tensor quantization for packed params?

supriyar replied:
Not atm, we only have per-row quantization support with float qparams.

raghuramank100 commented on torch/nn/quantized/modules/embedding.py:
```diff
@@ -0,0 +1,167 @@
+    #   |--- _packed_weight : Tensor representing weight of EmbeddingPackedParamsBase
+    #   |--- dtype : torch.dtype
+
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
```
Should we also have a field for bitwidth?

supriyar replied:
we can use the tensor dtype to determine the bitwidth, right? Currently it only supports 8bit, but once we add 4-bit  qtensor it should be encoded in the dtype.

raghuramank100 commented on torch/nn/quantized/modules/embedding.py:
```diff
@@ -0,0 +1,167 @@
+        super(Embedding, self).__init__()
+        self.num_embeddings = num_embeddings
+        self.embedding_dim = embedding_dim
+        self.sparse = sparse
```
For my understanding: What does self.sparse do ?

supriyar replied:
Currently it doesn't do anything for the quantized module. I'll remove it from here.
It is used in the float module for sparse gradients for weight tensor.

supriyar committed:
Update on "[quant] Add quantized Embedding module"


Summary:
Add quantized module in static quantization namespace. Embedding
quantization requires only weights to be quantized so it is static.
Internally it calls the embedding_bag_byte op with the offsets set corresponding to the
indices.

Future PR will move EmbeddingBag quantization from dynamic to static as well.

Test Plan:
python test/test_quantization.py test_embedding_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23547384](https://our.internmc.facebook.com/intern/diff/D23547384)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add quantized Embedding module"


Summary:
Add quantized module in static quantization namespace. Embedding
quantization requires only weights to be quantized so it is static.
Internally it calls the embedding_bag_byte op with the offsets set corresponding to the
indices.

Future PR will move EmbeddingBag quantization from dynamic to static as well.

Test Plan:
python test/test_quantization.py test_embedding_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23547384](https://our.internmc.facebook.com/intern/diff/D23547384)

[ghstack-poisoned]
